src/core/ruby_version.rs

```diff
@@ -1003,81 +1011,293 @@ macro_rules! get_stack_frame_2_5_0(
+            // Note: saturating_sub above already handles the < 0 case
+            if prefix_len > label_length {
+                prefix_len = label_length;
+            }
+
+            let profile_label = format!("{}{}", &label[..prefix_len], qualified);
+
+            if profile_label.is_empty() {
+                return String::new();
             }
-            let class_name = get_ruby_string(klass.classext.classpath as usize, source)?;
-            Ok(class_name.to_string())
+
+            // Get the prefix from label and concatenate with qualified_method_name
+            profile_label
         }
+    )
+);
 
+macro_rules! get_stack_frame_3_3_0(
+    () => (
         fn get_stack_frame<T>(
             iseq_struct: &rb_iseq_struct,
             cfp: &rb_control_frame_t,
             source: &T,
         ) -> Result<StackFrame> where T: ProcessMemory {
-            if iseq_struct.body == std::ptr::null_mut() {
-                return Err(format_err!("iseq body is null"));
-            }
-            let body: rb_iseq_constant_body = source.copy_struct(iseq_struct.body as usize)
-                .context("couldn't copy rb_iseq_constant_body")?;
-
+            //if iseq_struct.body == std::ptr::null_mut() {
+            //    return Err(format_err!("iseq body is null"));
+            //}
+            let cme = locate_method_entry(&cfp.ep, source)?;
+            let mut class_path = "".to_string();
+            let mut singleton = false;
+            let iseq = if cme != 0 {
+                let imemo: rb_method_entry_struct = source.copy_struct(cme).context(cme)?;
+                let method_type = source.copy(imemo.def as usize, 1).context(imemo.def as usize)?;
+                let method_type = method_type[0] & 0xf;
+                if method_type == 0 {
+                    (class_path, singleton) = get_classpath(cme, false, source)?;
+                    let method_def: rb_method_definition_struct = source.copy_struct(imemo.def as usize).context(imemo.def as usize)?;
+                    let iseq: rb_iseq_struct = unsafe {
+                        source.copy_struct(method_def.body.iseq.iseqptr as usize).context("")?
+                    };
+                    iseq
+                } else {
+                    *iseq_struct
+                }
+            } else {
+                *iseq_struct
+            };
+            let body: rb_iseq_constant_body =
+                source.copy_struct(iseq.body as usize)
+                    .context("couldn't copy rb_iseq_constant_body")?;
+            let frame_flag: usize = unsafe {
+                source.copy_struct(cfp.ep.offset(0) as usize).context(cfp.ep.offset(0) as usize)?
+            };
             let rstring: RString = source.copy_struct(body.location.label as usize)
                 .context("couldn't copy RString")?;
+
+            let mut method_name = "".to_string();
+            if body.local_iseq != std::ptr::null_mut() {
```

Comment:
We should be able to depend on this field being there most of the time, but to be safe we add this defensive check.

TODO add documentation link for why we do the indirection to local_iseq here.